File: 240102/두 방향 탈출 가능 여부 판별하기/determine-escapableness-with-2-ways.cpp
#include <iostream>
#define MAX_N 100
using namespace std;
int n,m;

int dirs[2][2] = {{1,0},{0,1}}; // 아래, 오른쪽
int graph[MAX_N][MAX_N];
bool visited[MAX_N][MAX_N];
int ans;

bool InRange(int x, int y){
    return 0 <= x && x < n && 0 <= y && y < m;
}

bool CanGo(int x, int y){
    // 범위 안에 있고, 방문하지 않았으며, 빈칸일 경우 갈 수 있음
    return InRange(x,y) && !visited[x][y] && graph[x][y];
}

void DFS(int x, int y){
    for(int i=0; i<2; i++){
        int nx = x+dirs[i][0], ny = y+dirs[i][1];
        if(CanGo(nx,ny)){
            visited[nx][ny] = 1;
            DFS(nx,ny);
        }
    }
}

int main(){
    cin >> n >> m;
    for(int i=0; i<n; i++){
        for(int j=0; j<m; j++){
            cin >> graph[i][j]; // 1: 빈칸, 0: 뱀
        }
    }

    visited[0][0] = true; // 시작칸 방문 처리
    DFS(0,0);
    
    if(visited[n-1][m-1]){
        ans = 1;
    }
    cout << ans;
    return 0;
}

File: 240127/두 방향 탈출 가능 여부 판별하기/determine-escapableness-with-2-ways.py
# 변수 선언 및 입력
n, m = tuple(map(int, input().split()))

graph = [
    # 빈칸:1, 뱀:0
    list(map(int, input().split()))
    for _ in range(n)
]

visited = [
    [0 for _ in range(m)] 
    for _ in range(n)
]
dirs = [[1,0],[0,1]] # 아래, 오른쪽

# 주어진 위치가 격자를 벗어나는지 여부를 반환
def InRange(x, y):
    return 0 <= x and x < n and 0 <= y and y < m

# 주어진 위치로 이동할 수 있는지 여부를 확인
def CanGo(x, y):
    return InRange(x,y) and not visited[x][y] and graph[x][y]

def dfs(x,y):
    for i in range(2):
        nx, ny = x+dirs[i][0], y+dirs[i][1]
        if CanGo(nx,ny):
            visited[nx][ny] = 1
            dfs(nx,ny)

r,c = 0,0
visited[r][c] = 1
dfs(r,c)

print(visited[n-1][m-1])